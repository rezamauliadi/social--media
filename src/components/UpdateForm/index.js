import React, { Component } from "react";
import PropTypes from "prop-types";

import { Button, Form, Message } from "semantic-ui-react";

import API from "src/api";

class UpdateForm extends Component {
  state = {
    id: 0,
    body: "",
    title: "",
    type: "",
    tempBody: "",
    tempTitle: "",
    invalid: false,
    loading: false,
    maxDummyId: {
      posts: 100,
      comments: 500
    }
  };

  changeValue = (_event, { name, value }) => {
    this.setState({ [name]: value });
  };

  getPayload = type => {
    const { id, body, title } = this.state;

    if (type === "posts") {
      const { parentId: userId } = this.props;
      return { id, title, body, userId };
    }

    const { parentId: postId, additionalInfo: email } = this.props;
    return {
      id,
      body,
      postId,
      email,
      name: title
    };
  };

  submitForm = async () => {
    this.setState({ loading: true, invalid: false });

    const { id, body, title, maxDummyId } = this.state;
    const { type } = this.props;

    if (!body || !title) {
      this.setState({ loading: false, invalid: true });
      return;
    }

    const payload = this.getPayload(type);
    if (id > maxDummyId[type]) {
      this.setState({ loading: false });
      this.props.afterSubmit(payload);
    } else {
      const endpoint = type === "posts" ? API.updatePost : API.updateComment;
      const item = await endpoint(payload, id);
      this.setState({ loading: false });
      this.props.afterSubmit(item);
    }
  };

  closeForm = () => {
    const { tempTitle, tempBody } = this.state;
    this.props.onClose(tempTitle, tempBody);
  };

  async componentDidMount() {
    const { id, title, body, type } = this.props;
    this.setState({
      id,
      title,
      body,
      type,
      tempBody: body,
      tempTitle: title
    });
  }

  render() {
    const { body, title, invalid, loading } = this.state;

    return (
      <Form warning={invalid}>
        <Form.Input
          placeholder="Give it a title!"
          name="title"
          value={title}
          onChange={this.changeValue}
        />
        <Form.TextArea
          placeholder="Write your post!"
          name="body"
          value={body}
          rows="5"
          onChange={this.changeValue}
        />
        <Message size="mini" warning>
          It will be great if you put some words in the post and give it a cool
          title.
        </Message>
        <Button
          size="tiny"
          icon="send"
          primary
          loading={loading}
          disabled={loading}
          onClick={() => this.submitForm()}
        />
        <Button
          size="tiny"
          icon="cancel"
          color="orange"
          disabled={loading}
          onClick={() => this.closeForm()}
        />
      </Form>
    );
  }
}

UpdateForm.propTypes = {
  id: PropTypes.number,
  parentId: PropTypes.number,
  type: PropTypes.string,
  title: PropTypes.string,
  body: PropTypes.string,
  additionalInfo: PropTypes.string,
  onClose: PropTypes.func,
  afterSubmit: PropTypes.func
};

export default UpdateForm;
