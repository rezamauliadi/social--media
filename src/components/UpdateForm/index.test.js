import React from "react";
import { shallow } from "enzyme";

import UpdateForm from "./index.js";
import API from "src/api";

describe("Component: UpdateForm", () => {
  const fnMock = jest.fn();
  const props = {
    id: 1,
    parentId: 1,
    type: "posts",
    title: "foo",
    body: "bar",
    additionalInfo: "user@email.com",
    onClose: fnMock,
    afterSubmit: fnMock
  };

  it("should match snapshot", () => {
    const component = shallow(<UpdateForm {...props} />);
    expect(component).toMatchSnapshot();
  });

  describe("button click", () => {
    it("should call onClose when close button clicked", () => {
      const component = shallow(<UpdateForm {...props} />);
      const instance = component.instance();

      jest.spyOn(instance, "closeForm");
      jest.spyOn(instance.props, "onClose");
      component
        .find("Button")
        .at(1)
        .simulate("click");

      expect(instance.closeForm).toHaveBeenCalled();
      expect(instance.props.onClose).toHaveBeenCalled();
    });

    it("should do update when update button clicked", () => {
      const component = shallow(<UpdateForm {...props} />);
      const instance = component.instance();

      jest.spyOn(instance, "submitForm");
      jest.spyOn(instance, "getPayload");
      component
        .find("Button")
        .at(0)
        .simulate("click");

      expect(instance.submitForm).toHaveBeenCalled();
      expect(instance.getPayload).toHaveBeenCalled();
    });
  });

  describe("changeValue", () => {
    it("should set correct state when changeValue", () => {
      const component = shallow(<UpdateForm {...props} />);
      const instance = component.instance();
      const data = {
        name: "title",
        value: "foo"
      };

      instance.changeValue({}, data);
      expect(component.state("title")).toBe("foo");
    });
  });

  describe("getPayload", () => {
    it("should return correct payload (type: posts)", () => {
      const component = shallow(<UpdateForm {...props} />);
      const instance = component.instance();
      const expected = {
        id: 1,
        title: "foo",
        body: "bar",
        userId: 1
      };

      const payload = instance.getPayload("posts");
      expect(payload).toEqual(expected);
    });

    it("should return correct payload (type: comments)", () => {
      const component = shallow(<UpdateForm {...props} />);
      const instance = component.instance();
      const expected = {
        id: 1,
        body: "bar",
        email: "user@email.com",
        name: "foo",
        postId: 1
      };

      const payload = instance.getPayload("comments");
      expect(payload).toEqual(expected);
    });
  });

  describe("submitForm", () => {
    it("should return on submitForm when no body or title", done => {
      const component = shallow(<UpdateForm {...props} />);
      const instance = component.instance();

      jest.spyOn(API, "updatePost").mockImplementation(() => {});
      jest.spyOn(instance, "getPayload");
      component.setState({ body: "", title: "" });

      instance.submitForm();

      process.nextTick(() => {
        expect(instance.getPayload).not.toHaveBeenCalled();
        done();
      });
    });

    it("should return payload on submitForm when id exceed maxDummy", done => {
      const component = shallow(<UpdateForm {...props} />);
      const instance = component.instance();

      jest.spyOn(instance.props, "afterSubmit");
      component.setState({ id: 101 });

      instance.submitForm();

      process.nextTick(() => {
        expect(instance.props.afterSubmit).toHaveBeenCalled();
        done();
      });
    });

    it("should do submitForm (type: posts)", done => {
      const component = shallow(<UpdateForm {...props} />);
      const instance = component.instance();

      jest.spyOn(API, "updatePost").mockImplementation(() => {});
      jest.spyOn(instance.props, "afterSubmit");
      jest.spyOn(instance, "getPayload");

      instance.submitForm();

      process.nextTick(() => {
        expect(instance.getPayload).toHaveBeenCalled();
        expect(instance.props.afterSubmit).toHaveBeenCalled();
        done();
      });
    });

    it("should do submitForm (type: comments)", done => {
      const newProps = { ...props };
      newProps.type = "comments";
      const component = shallow(<UpdateForm {...newProps} />);
      const instance = component.instance();

      jest.spyOn(API, "updateComment").mockImplementation(() => {});
      jest.spyOn(instance.props, "afterSubmit");
      jest.spyOn(instance, "getPayload");

      instance.submitForm();

      process.nextTick(() => {
        expect(instance.getPayload).toHaveBeenCalled();
        expect(instance.props.afterSubmit).toHaveBeenCalled();
        done();
      });
    });
  });
});
