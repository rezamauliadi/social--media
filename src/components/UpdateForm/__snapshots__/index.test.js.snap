// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: UpdateForm should match snapshot 1`] = `
<Form
  as="form"
  warning={false}
>
  <FormInput
    as={[Function]}
    control={[Function]}
    name="title"
    onChange={[Function]}
    placeholder="Give it a title!"
    value="foo"
  />
  <FormTextArea
    as={[Function]}
    control={[Function]}
    name="body"
    onChange={[Function]}
    placeholder="Write your post!"
    rows="5"
    value="bar"
  />
  <Message
    size="mini"
    warning={true}
  >
    It will be great if you put some words in the post and give it a cool title.
  </Message>
  <Button
    as="button"
    disabled={false}
    icon="send"
    loading={false}
    onClick={[Function]}
    primary={true}
    size="tiny"
  />
  <Button
    as="button"
    color="orange"
    disabled={false}
    icon="cancel"
    onClick={[Function]}
    size="tiny"
  />
</Form>
`;
